File: C++/ex20/lru_cache_solution.cpp
/**
 * Exercise 20: LRU (Least Recently Used) Cache
 * Difficulty: ***** (5/5)
 * 
 * Implement an LRU Cache that supports get and put operations.
 * When the cache reaches capacity, evict the least recently used item.
 * 
 * Input:
 * - First line: cache capacity
 * - Second line: number of operations
 * - Next n lines: operations ("get key" or "put key value")
 * 
 * Output:
 * - For each get operation, print the value (or -1 if not found)
 */

#include <bits/stdc++.h>
using namespace std;

class LRUCache {
private:
    int capacity;
    unordered_map<int, pair<int, list<int>::iterator>> cache;
    list<int> lruList;
    
public:
    LRUCache(int cap) : capacity(cap) {}
    
    int get(int key) {
        if (cache.find(key) == cache.end()) {
            return -1;
        }
        
        // Move to front (most recently used)
        lruList.erase(cache[key].second);
        lruList.push_front(key);
        cache[key].second = lruList.begin();
        
        return cache[key].first;
    }
    
    void put(int key, int value) {
        if (cache.find(key) != cache.end()) {
            // Key already exists, update value and move to front
            lruList.erase(cache[key].second);
            lruList.push_front(key);
            cache[key] = {value, lruList.begin()};
        } else {
            // New key
            if (cache.size() == capacity) {
                // Evict least recently used
                int lru = lruList.back();
                lruList.pop_back();
                cache.erase(lru);
            }
            
            lruList.push_front(key);
            cache[key] = {value, lruList.begin()};
        }
    }
};

int main() {
    int capacity;
    cin >> capacity;
    
    LRUCache cache(capacity);
    int n;
    cin >> n;
    
    for (int i = 0; i < n; i++) {
        string cmd;
        cin >> cmd;
        
        if (cmd == "get") {
            int key;
            cin >> key;
            cout << cache.get(key) << "\n";
        } else if (cmd == "put") {
            int key, value;
            cin >> key >> value;
            cache.put(key, value);
        }
    }
    
    return 0;
}

File: C#/ex36/lru_cache_solution.cs
/**
 * Exercise 36: LRU Cache
 * Difficulty: ***** (5/5)
 * 
 * Solution using Dictionary and LinkedList for O(1) get and put operations.
 */

using System;
using System.Collections.Generic;

class CacheItem
{
    public int Key { get; set; }
    public int Value { get; set; }

    public CacheItem(int key, int value)
    {
        Key = key;
        Value = value;
    }
}

class LRUCache
{
    private int capacity;
    private Dictionary<int, LinkedListNode<CacheItem>> cache;
    private LinkedList<CacheItem> lruList;

    public LRUCache(int capacity)
    {
        this.capacity = capacity;
        cache = new Dictionary<int, LinkedListNode<CacheItem>>();
        lruList = new LinkedList<CacheItem>();
    }

    public int Get(int key)
    {
        if (!cache.ContainsKey(key))
        {
            return -1;
        }

        // Move accessed item to front (most recently used)
        LinkedListNode<CacheItem> node = cache[key];
        lruList.Remove(node);
        lruList.AddFirst(node);

        return node.Value.Value;
    }

    public void Put(int key, int value)
    {
        if (cache.ContainsKey(key))
        {
            // Update existing key
            LinkedListNode<CacheItem> node = cache[key];
            node.Value.Value = value;
            
            // Move to front (most recently used)
            lruList.Remove(node);
            lruList.AddFirst(node);
        }
        else
        {
            // Check capacity
            if (cache.Count >= capacity)
            {
                // Remove least recently used (last item)
                LinkedListNode<CacheItem> lru = lruList.Last;
                cache.Remove(lru.Value.Key);
                lruList.RemoveLast();
            }

            // Add new item to front
            CacheItem newItem = new CacheItem(key, value);
            LinkedListNode<CacheItem> newNode = new LinkedListNode<CacheItem>(newItem);
            lruList.AddFirst(newNode);
            cache[key] = newNode;
        }
    }
}

class Program
{
    static void Main()
    {
        int capacity = int.Parse(Console.ReadLine());
        int n = int.Parse(Console.ReadLine());
        LRUCache cache = new LRUCache(capacity);

        for (int i = 0; i < n; i++)
        {
            string[] input = Console.ReadLine().Split();
            string operation = input[0];

            if (operation == "get")
            {
                int key = int.Parse(input[1]);
                Console.WriteLine(cache.Get(key));
            }
            else if (operation == "put")
            {
                int key = int.Parse(input[1]);
                int value = int.Parse(input[2]);
                cache.Put(key, value);
            }
        }
    }
}
